zcash_client_sqlite/src/wallet.rs

```diff
@@ -2020,9 +1849,31 @@ pub(crate) fn truncate_to_height<P: consensus::Parameters>(
         named_params![":new_end_height": u32::from(block_height + 1)],
     )?;
 
-    // If we're removing scanned blocks, we need to truncate the note commitment tree, un-mine
-    // transactions, and remove received transparent outputs and affected block records from the
-    // database.
+    // Mark transparent utxos as un-mined. Since the TXO is now not mined, it would ideally be
+    // considered to have been returned to the mempool; it _might_ be spendable in this state, but
+    // we must also set its max_observed_unspent_height field to NULL because the transaction may
+    // be rendered entirely invalid by a reorg that alters anchor(s) used in constructing shielded
+    // spends in the transaction.
+    conn.execute(
+        "UPDATE transparent_received_outputs
+         SET max_observed_unspent_height = CASE WHEN tx.mined_height <= :height THEN :height ELSE NULL END
+         FROM transactions tx
+         WHERE tx.id_tx = transaction_id
+         AND max_observed_unspent_height > :height",
+        named_params![":height": u32::from(block_height)],
+    )?;
+
+    // Un-mine transactions. This must be done outside of the last_scanned_height check because
+    // transaction entries may be created as a consequence of receiving transparent TXOs.
+    conn.execute(
+        "UPDATE transactions
+         SET block = NULL, mined_height = NULL, tx_index = NULL
+         WHERE mined_height > :height",
+        named_params![":height": u32::from(block_height)],
+    )?;
+
+    // If we're removing scanned blocks, we need to truncate the note commitment tree and remove
+    // affected block records from the database.
     if block_height < last_scanned_height {
         // Truncate the note commitment trees
         let mut wdb = WalletDb {
```

Comment:
Aside: I don't understand why we're creating a new `WalletDb` wrapper when we could instead pass in `wdb: &mut WalletDb<SqlTransaction<'_>, P>`.

If I understand correctly, the point of the `SqlTransaction` wrapper is to mark that it was constructed by `WalletDb::transactionally`. This does several useful things:
* It guarantees that the `rusqlite::Transaction` has `DEFERRED` behaviour, because `WalletDb::transactionally` calls `self.conn.transaction()` which is implemented to call `rusqlite::Transaction::new(conn, TransactionBehavior::Deferred)`. The meaning of `DEFERRED` is that the transaction rolls back unless `commit()` is called explicitly (given that we never call `rusqlite::Connection::set_drop_behavior`). `WalletDb::transactionally` calls `commit()` iff the closure returns `Ok(_)`, which is the proper behaviour of a Rust database API that we should always want.
* It also guarantees that we can't accidentally commit earlier than we intended to. SQLite doesn't support nested transactions (it does support [savepoints](https://www.sqlite.org/lang_savepoint.html) which can be used to simulate them, but we don't use that). If some called function `f` that needs to be transactional were responsible for calling `commit()`, that would be a hazard because we couldn't reuse `f` in a context that also needs to be transactional: `f` would call `commit()` too early. This is avoided with the `transactionally` API because the transactional code never needs to call `commit()`.
  * Note that the fact that `rusqlite::Connection::transaction` takes a `&mut` reference does not prevent this hazard: it prevents *creating* nested transactions at the SQLite API layer (if you don't use the `Transaction::new_unchecked` escape hatch), but it can't prevent user code from committing too early.
* It enforces single-threaded use of the connection when it is in a transaction, which (even if SQLite didn't internally use locks in a way that effectively prevents any useful concurrency) is necessary for the transaction API to even make sense: you have to do all the things *and then* call `commit()`.

Here we don't call `commit()` in `truncate_to_height`, which is correct because we take a `rusqlite::Transaction` as argument, but that means:
* we don't know what the transaction behaviour is set to without tracing where `conn` came from;
* we don't have any guarantee that the caller has taken responsibility for calling `commit()`.

Please let's refactor (not in this PR) so that we never need to explicitly create a `WalletDb` wrapper like this outside `WalletDb::transactionally`.

Reply:
`SqlTransaction` was added to the codebase much later on, to deal with some problems around how to implement the `WalletCommitmentTrees` trait. I can't recall all the problems that I ran into, but I remember spending several days and multiple pairings with @str4d figuring out an approach that could work here; there's some problem if the connection type is the same in `WalletWrite` as in `WalletCommitmentTrees.`

Reply:
Filed #1435.